cura/Machines/MachineNode.py

```diff
@@ -0,0 +1,172 @@
+
+        container_registry = ContainerRegistry.getInstance()
+        try:
+            my_metadata = container_registry.findContainersMetadata(id = container_id)[0]
+        except IndexError:
+            Logger.log("Unable to find metadata for container %s", container_id)
+            my_metadata = {}
+        # Some of the metadata is cached upon construction here.
+        # ONLY DO THAT FOR METADATA THAT DOESN'T CHANGE DURING RUNTIME!
+        # Otherwise you need to keep it up-to-date during runtime.
+        self.has_materials = parseBool(my_metadata.get("has_materials", "true"))
+        self.has_variants = parseBool(my_metadata.get("has_variants", "false"))
+        self.has_machine_materials = parseBool(my_metadata.get("has_machine_materials", "false"))
+        self.has_machine_quality = parseBool(my_metadata.get("has_machine_quality", "false"))
+        self.quality_definition = my_metadata.get("quality_definition", container_id)
+        self.exclude_materials = my_metadata.get("exclude_materials", [])
+        self.preferred_variant_name = my_metadata.get("preferred_variant_name", "")
+        self.preferred_material = my_metadata.get("preferred_material", "")
+        self.preferred_quality_type = my_metadata.get("preferred_quality_type", "")
+
+        self._loadAll()
+
+    ##  Get the available quality groups for this machine.
+    #
+    #   This returns all quality groups, regardless of whether they are
+    #   available to the combination of extruders or not. On the resulting
+    #   quality groups, the is_available property is set to indicate whether the
+    #   quality group can be selected according to the combination of extruders
+    #   in the parameters.
+    #   \param variant_names The names of the variants loaded in each extruder.
+    #   \param material_bases The base file names of the materials loaded in
+    #   each extruder.
+    #   \param extruder_enabled Whether or not the extruders are enabled. This
+    #   allows the function to set the is_available properly.
+    #   \return For each available quality type, a QualityGroup instance.
+    def getQualityGroups(self, variant_names: List[str], material_bases: List[str], extruder_enabled: List[bool]) -> Dict[str, QualityGroup]:
+        if len(variant_names) != len(material_bases) or len(variant_names) != len(extruder_enabled):
+            Logger.log("e", "The number of extruders in the list of variants (" + str(len(variant_names)) + ") is not equal to the number of extruders in the list of materials (" + str(len(material_bases)) + ") or the list of enabled extruders (" + str(len(extruder_enabled)) + ").")
+            return {}
+        # For each extruder, find which quality profiles are available. Later we'll intersect the quality types.
+        qualities_per_type_per_extruder = [{} for _ in range(len(variant_names))]  # type: List[Dict[str, QualityNode]]
```

Comment:
Minor: you can do this: `[{}] * len(variant_names)` so you don't need to iterate. This works with 0, too.

Reply:
Done: 35907e52285c9f46e3c030a516c1e7ccb86d3d7e